### PR TITLE
Eliminate need for CompartmentalModel.transition_bwd()

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2426 

This follows ideas from #2513 to replace the `.transition_bwd()` method with a conditioned call of the `.transition_fwd()` method, significantly reducing user effort to create a new model. The additional bit of logic to reverse the flow computation is implemented in a new `.compute_flows()` method that by default assumes a linear flow terminating in "R", but can be overridden by users to perform arbitrary computations. Only one such override is needed by our current examples.

The new `.transition()` method is now called under more interpretations, so various logic needs to be generalized:
- `t < self.duration` now also needs to check for `slice` instances
- We now need to use `ExtendedBinomial` everywhere, not only in `.transition_bwd()`
- The SMC wrapper now needs to provide approximate variables like `I_approx`.

## Tested
- [x] refactoring covered by existing tests
- [x] manually ran a biggish example and verified sane posterior

### CODE REVIEW

File: pyro/contrib/epidemiology/models.py [hunk 1]
```diff
@@ -771,38 +623,23 @@ def transition_fwd(self, params, state, t):
...
+        mask_t = self.mask[t] if t_is_observed else False
+        data_t = self.data[t] if t_is_observed else None
         pyro.sample("obs_{}".format(t),
                     dist.Delta(state["O"]).mask(mask_t),
                     obs=data_t)
 
-    def transition_bwd(self, params, prev, curr, t):
-        R0, tau, rho = params
-
+    def compute_flows(self, prev, curr, t):
```
Comment (fritzo) [inline pyro/contrib/epidemiology/models.py:633]: This is the one model where we need a custom `.compute_flows()` method.

File: pyro/contrib/epidemiology/compartmental.py [hunk 2]
```diff
@@ -172,50 +172,68 @@ def initialize(self, params):
...
 
         This inputs a current ``state`` and stochastically updates that
         state in-place.
 
-        Note that this method is called under two different interpretations.
+        Note that this method is called under multiple different
+        interpretations, including batched and vectorized interpretations.
         During :meth:`generate` this is called to generate a single sample.
-        During :meth:`predict` thsi is called to forecast a batch of samples,
+        During :meth:`heuristic` this is called to generate a batch of sample
```
Comment (martinjankowiak) [inline pyro/contrib/epidemiology/compartmental.py:185]: typo: samples

File: pyro/contrib/epidemiology/compartmental.py [hunk 3]
```diff
@@ -172,50 +172,68 @@ def initialize(self, params):
...
+        for SMC.  During :meth:`fit` this is called both in vectorized form
+        (vectorizing over time) and insequential form (for a single time step);
```
Comment (martinjankowiak) [inline pyro/contrib/epidemiology/compartmental.py:187]: typo: insequential

File: pyro/contrib/epidemiology/models.py [hunk 4]
```diff
@@ -717,6 +564,11 @@ class SparseSIRModel(CompartmentalModel):
     (when ``mask[t] == True``) ``O`` must exactly match the provided data;
     between observed times ``O`` stochastically imputes the provided data.
 
+    This model demonstrates how to implement a custom :meth:`compute_flows`
+    method. A custom method is needed in this model because inhabitants of the
+    ``S`` compartment can transition to both the ``I`` and ``o`` compartments,
```
Comment (martinjankowiak) [inline pyro/contrib/epidemiology/models.py:569]: typo: o versus O

**Pull request merged**